### PR TITLE
Split execution and planning (refactoring)

### PR DESCRIPTION
**Pull request by cecton**
## List of changes in this PR

- ~~**Renamed models.rs to query_plan.rs:** these models were meant to work on for the `QueryPlan` and I would like to make it more explicit~~ [edit: revert for now as it makes the review harder]
- **Use String instead of Query in Request:** the query there was fine but involved parsing. So Request will keep a simple string but a `Query` can be parsed from this string later on
- **Naive introspection knits:** takes a query string in parameter instead of an entire request + return directly a response instead of a Value that we convert to a response anyway.
- **Moved/Added validation functions to PlanNode:** all the validation functions require the plan so I moved them there.
- **Moved plan execution to PlanNode:** the user doesn't need to customize that.
- **Added a wrapper to the execute() PlanNode function:** this makes the public API of PlanNode easier.

## Most important changes

- Remove Fetcher from FederatedGraph
- Rename FederatedGraph to Router
- Added 2 new traits: Router and Route
  - A "Router" is an object that creates route
  - A "Route" is an object that you execute to create the ResponseStream
- **Moved Router/Route/FederatedGraph's logic to apollo-router/apollo_router.rs:** this now becomes the "default implementation of a router". The user **will** be able to replace those parts.

Related to #106 #62 #100

### CODE REVIEW

**Comment by Geal** [comment]
is there any way this could be split into multiple PRs? Especially considering we'll make decisions next week about which parts can or cannot be modified by users, so in the meantime we could review and merge the less risky parts

**Comment by Geal** [comment]
it's fine if those PRs build up on each other, btw. Just, it's harder to review a PR if it changes one file and moves/renames that file at the same time

**Comment by cecton** [comment]
> it's fine if those PRs build up on each other, btw. Just, it's harder to review a PR if it changes one file and moves/renames that file at the same time

You're right, I'll move back what I can.

> is there any way this could be split into multiple PRs? 

This PR is actually already a pre-emptive PR for other changes xD Though I could probably separate on the points I mentioned on the ticket itself... :thinking: but that will be a lot of work. Also because I don't want to commit not-working things into main. I will rename the file that has been renamed so it's easier to review.

Let me know if you still have issues reviewing

**Comment by cecton** [comment]
It worked for model.rs :raised_hands: This is ready for review again @Geal @o0Ignition0o @BrynCooke

File: crates/apollo-router-core/src/naive_introspection.rs
Comment (Geal) [inline crates/apollo-router-core/src/naive_introspection.rs:59]: that could be a method in the response builder, like `with_data(value)`

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (Geal) [inline crates/apollo-router-core/src/query_planner/model.rs:88]: do we really need to do that? Since the query plan was generated from the schema, and the service registry was generated from the schema too, we're sure they will all be there. Or are there other cases that we should prepare for? Like circuit breaking for a failing service?

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (Geal) [inline crates/apollo-router-core/src/query_planner/model.rs:93]: not sure we need to collect here, since we're going through the entire list anyway, we could check in the service registry for each of them, it's just a hashmap lookup

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (Geal) [inline crates/apollo-router-core/src/query_planner/model.rs:149]: it makes much more sense to have this here :+1:

File: crates/apollo-router/src/apollo_router.rs
Comment (Geal) [inline crates/apollo-router/src/apollo_router.rs:36]: not sure about calling it `create_route` and return an `ApolloRoute`. Can we have a word around "query" instead? Since this is handling one query

File: crates/apollo-router/src/http_server_factory.rs
Comment (Geal) [inline crates/apollo-router/src/http_server_factory.rs:27]: why is there a `'static` here? From what I see it's definitely not `'static`

**Review by Geal — CHANGES_REQUESTED** [review]

File: crates/apollo-router/src/http_server_factory.rs
Comment (cecton) [reply crates/apollo-router/src/http_server_factory.rs:27]: :thinking: ...

Just to be clear we understand each other. You do know `'static` in trait boundary is not at all the same than `'static` in lifetimes?

File: crates/apollo-router/src/apollo_router.rs
Comment (cecton) [reply crates/apollo-router/src/apollo_router.rs:36]: It's not "handling" the whole query. It's creating a `Route` instance which will "handle" (execute) the query.

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (cecton) [reply crates/apollo-router-core/src/query_planner/model.rs:93]: I did not change this in this PR. I plan to refactor the `ServiceRegistry` in a future PR to avoid making this one heavier.

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (cecton) [reply crates/apollo-router-core/src/query_planner/model.rs:88]: Again this is old code. You're probably right that when the query will be parsed and validated against schema we won't need this anymore.

File: crates/apollo-router-core/src/naive_introspection.rs
Comment (cecton) [reply crates/apollo-router-core/src/naive_introspection.rs]: Good catch! Fixed in 1fbffe5 cc @o0Ignition0o (yet another small knit with naive introspection xD)

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (cecton) [reply crates/apollo-router-core/src/query_planner/model.rs:149]: ikr! :laughing:

File: crates/apollo-router/src/http_server_factory.rs
Comment (cecton) [reply crates/apollo-router/src/http_server_factory.rs:27]: I checked and we do a tokio::spawn at some point with that. So it needs to be owned (`'static`). See the `'static` [here](https://docs.rs/async-std/1.6.1/async_std/task/fn.spawn.html)

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (cecton) [reply crates/apollo-router-core/src/query_planner/model.rs:88]: (Do you mind creating an issue from this comment? :grin: )

File: crates/apollo-router/src/apollo_router.rs
Comment (cecton) [reply crates/apollo-router/src/apollo_router.rs:36]: Naming is hard but I like the `Router`'s `create_route()` literally returning a newly created `Route`. If others have suggestions I will change to whatever the consensus go.

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (o0Ignition0o) [reply crates/apollo-router-core/src/query_planner/model.rs:88]: the [bridge planner](https://github.com/apollographql/federation/blob/main/router-bridge/js-src/plan.ts#L18) calls functions that validate the [operation against the schema](https://github.com/apollographql/federation/blob/3b2f486e876f8da8a66e3e94bfae8d8db0c2d2cc/internals-js/src/operations.ts#L1257) as well as the fragments within, before building the query plan.

It might be worth a test though.

Comment (Geal) [reply crates/apollo-router-core/src/query_planner/model.rs:88]: https://github.com/apollographql/router/issues/177

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (Geal) [reply crates/apollo-router-core/src/query_planner/model.rs:93]: this can be collected in a `HashSet` after the filter. In the most common case it will be empty, instead of allocating for all services every time

File: crates/apollo-router/src/apollo_router.rs
Comment (Geal) [reply crates/apollo-router/src/apollo_router.rs:36]: maybe an Execution or something like that? In databases what comes after the query planner is the executor, in charge of executing the steps.

File: crates/apollo-router/src/http_server_factory.rs
Comment (Geal) [reply crates/apollo-router/src/http_server_factory.rs:27]: ooof right, my mistake :sweat_smile:

File: crates/apollo-router/src/apollo_router.rs
Comment (cecton) [reply crates/apollo-router/src/apollo_router.rs:36]: I talked to @o0Ignition0o and I realized that the naming of Router and Route wasn't as obvious to me as it is to others :sweat: I'll try to explain the reasoning better:

> [...] graph router for a federated graph.

Apollo Router is a "graph router". What does a router do? It routes things. So the core of Apollo Router is to be a "router" (for graphs).

So what does a router do? It creates a path of execution, aka "route". The router create a specific "route" for a specific request which can then be executed.

> Rename FederatedGraph to Router

The whole point of the PR is to split the process of creation of this route from the execution of this route.

FederatedGraph => takes a request, query the services, return a response stream
Router => takes a request, make a route
Route => can be executed to get the response stream

Comment (cecton) [reply crates/apollo-router/src/apollo_router.rs:36]: (So the function on Router is called create_route because it creates a new route. While the function Route is called execute because it actually execute the process)

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (cecton) [reply crates/apollo-router-core/src/query_planner/model.rs:94]: Right!! Thanks! 80286fc

File: crates/apollo-router/src/apollo_router.rs
Comment (o0Ignition0o) [reply crates/apollo-router/src/apollo_router.rs:36]: I think the truth is the router isn't a router, It's a sequencer or an orchestrator or something like that. What it means is we're not actually routing queries, but Building an execution plan, and running it. This is IMO the reason why I was confused with ApolloRouter and ApolloRoute, also the apollo prefix doesn't make much sense here imo.

now it looks like an ApolloRoute executes a plan against uh services, and it passes _a lot_ of things to it as parameters. I think there's room for refactoring, but I don't know how exactly. I'll try to let you know if I can think of anything. The short version is that imo query_plan.execute() has too big parameters, so we might be able to move things around and come up with a more elegant set of functions

File: crates/apollo-router-core/src/traits.rs
Comment (BrynCooke) [inline crates/apollo-router-core/src/traits.rs:58]: Be nice to explain in the function doc why `ResponseStream` is returned as the error part of the `Result`, it's GraphQL and all that.

File: crates/apollo-router-core/src/traits.rs
Comment (BrynCooke) [inline crates/apollo-router-core/src/traits.rs:59]: I assume this is eventually going to allow us to support caching the preparation phase of a query. 

Possible renames:
`create_route` -> `prepare`
`Route` -> `PreparedQuery`

Also I would expect `execute` to take `request` as a parameter and not store it in `ApolloRoute`.

**Review by BrynCooke — COMMENTED** [review]
Generally looks great.

Thank you again for making all the commits digestible in their own right. I would not have been able to review without this.

Only real query is around the two traits `Router` and `Route`.

File: crates/apollo-router-core/src/query_planner/model.rs
Comment (cecton) [reply crates/apollo-router-core/src/query_planner/model.rs]: fine by me  (note that this function is **not** public)

Fixed in 9161f4e

File: crates/apollo-router-core/src/traits.rs
Comment (cecton) [reply crates/apollo-router-core/src/traits.rs:58]: Good point! Fixed in ddcefe2829b9841e0d04398613e74f46110f4e4c

I'm a bit in doubt if this should be a ResponseStream or a Response (single response object)... :thinking: It doesn't seem likely that erroring will result in a stream of errors. Or maybe it will? What do you think?

File: crates/apollo-router-core/src/traits.rs
Comment (cecton) [reply crates/apollo-router-core/src/traits.rs:63]: > I assume this is eventually going to allow us to support caching the preparation phase of a query.

:100: 

> Also I would expect execute to take request as a parameter and not store it in ApolloRoute.

I think you're :100: right. The object itself could be cached and the variables of the request may vary.

> Possible renames:
> create_route -> prepare
> Route -> PreparedQuery

I was so proud to find the name `Router -> create_route() -> Route -> execute()` :disappointed:  But I realized during the review that all of you got very confused with that and as I'm not really a graphql user myself I will refer to your judgment of you all to see what is the better name for these. cc @o0Ignition0o @Geal

Comment (o0Ignition0o) [reply crates/apollo-router-core/src/traits.rs:63]: 😅 I think these bits will come together once query_plan and query_plan.execute() do!

File: crates/apollo-router/src/apollo_router.rs
Comment (cecton) [reply crates/apollo-router/src/apollo_router.rs:36]: > the apollo prefix doesn't make much sense here imo.

The Apollo prefix is to differentiate the core Router trait from the struct provided by Apollo (the company) for general purpose.

> it looks like an ApolloRoute [...] it passes a lot of things to it as parameters.

A lot of things are passed in order to ensure the object will have (shared) ownership of the resources it requires to execute the process: the query plan itself, the service registry, the schema and a cached parsed query.

The request object though should be provided in argument as @BrynCooke mention as this might holds variables specific to a particular request.

> ApolloRoute executes a plan against uh services

Not services. A request. But that will be more obvious when I move request to the function arguments.

> The short version is that imo query_plan.execute() has too big parameters, so we might be able to move things around and come up with a more elegant set of functions

These are all immutable Arcs (low in memory usage) and meant to be used in scope of the request execution. Everything you need to be able to execute the query should be there. It's also important that these are internals to the Route itself as it can be customized differently by a different kind of Route.

Of course it's possible to make things simple: remove the lib, keep only the bin, remove all the traits. But my goal is to make this extensible to the user. Traits are a good way to make things extensible.

Comment (Geal) [reply crates/apollo-router/src/apollo_router.rs:36]: It's still very confusing for me. Router is an overloaded word, so we try to match it with a common meaning (the first one for me is a network component). The project's name is Apollo router, and it defines a HTTP server, available as library or executable, that receives multiple graphql queries over connections, and executes them. This a struct also called ApolloRouter, that does not receives HTTP requests directly, but we can pass it a request and it can return a future of Response, but actually it's not a HTTP request handler or [service](https://docs.rs/tower-service/0.3.1/tower_service/trait.Service.html), because it can return a route?

> It creates a path of execution, aka "route"

That is not evident, unfortunately.

I feel the issue is that we're confusing internal implementation details with user facing API.
If I'm a user that wants to build a custom router, with my own network layer (example: one endpoint receives graphql, others my own app). I need a [Service](https://docs.rs/hyper/0.14.15/hyper/service/trait.Service.html) I can call if I see graphql requests, maybe add middlewares. I give that service a request and it handles executing and returning a response, be it an introspection query or a larger query.

If I'm a user that wants to build a custom service, maybe to modify the query plan, filter info from introspection, then I need a structure that can:
- given a query, give me an introspection response
- given a query, give me a query plan
- given a query plan, executes the plan and gives me the response

That structure would let me handle the response afterwards. For that, the words "query planner", "orchestrator", "executor" work better, I think

Comment (Geal) [reply crates/apollo-router/src/apollo_router.rs:36]: > The Apollo prefix is to differentiate the core Router trait from the struct provided by Apollo (the company) for general purpose.

The Apollo router is the Router :wink:

File: crates/apollo-router-core/src/traits.rs
Comment (BrynCooke) [reply crates/apollo-router-core/src/traits.rs:58]: I think it's nice as is (ResponseStream), the caller doesn't have to deal with different types depending on if it was an error or not.

File: crates/apollo-router/src/apollo_router.rs
Comment (cecton) [reply crates/apollo-router/src/apollo_router.rs:36]: > The Apollo router is the Router :wink: 

And The Router answers to all queries...... :wink:

Comment (cecton) [reply crates/apollo-router/src/apollo_router.rs:36]: > The project's name is Apollo router, and it defines a HTTP server, available as library or executable, that receives multiple graphql queries over connections, and executes them. 

According to the README:
> The **Apollo Router** is a configurable, high-performance **graph router** for a [federated graph](https://www.apollographql.com/docs/federation/).

So it's actually a **graph** router. 

> This a struct also called ApolloRouter, that does not receives HTTP requests directly, but we can pass it a request and it can return a future of Response

Yes!! Almost. You pass it a `graphql::Request` (not http request) and it returns a `graphql::Route` in case of success or a `graphql::ResponseStream`.

> but actually it's not a HTTP request handler or service, because it can return a route?

Exactly. It is agnostic of the HTTP layer. It's above http, it's a GraphQL thingy. It doesn't return any route, it returns a graphql::Route.

> > It creates a path of execution, aka "route"
>
> That is not evident, unfortunately.

imo: a router does routing.

tr.v. rout·ed, rout·ing, routes
1. To send or forward by a specific route. See Synonyms at send.
2. To schedule the order of (a sequence of procedures).

@o0Ignition0o: this is starting to sounds a lot like "It's a sequencer or an orchestrator or something like that.")
@Geal: "the words "query planner", "orchestrator", "executor" work better, I think"

I don't know I'm not a native English speaker but a router doing routing by creating routes that can be executed still sounds good to me.

> I feel the issue is that we're confusing internal implementation details with user facing API.

I made this PR to improve the user facing API. They can now implement their own implementation of Router or Route. This was not possible before. They can make an `ApolloRouter<MyCustomRoute>` or a `MyCustomRouter<ApolloRoute>`. So they can hook right before the query execution itself or right before the route preparation itself.

(API might not be complete for `ApolloRouter<MyCustomRoute>` but that's the general idea)

Comment (Geal) [reply crates/apollo-router/src/apollo_router.rs:36]: It's starting to feel bikesheddy, and that should not block this PR, especially if we want to modify those parts of the code and avoid conflicts.
Let's go with those words for now. Can you open an issue and tag people to discuss how we should name that part and revisit that later?

File: crates/apollo-router-core/src/traits.rs
Comment (cecton) [reply crates/apollo-router-core/src/traits.rs:64]: Fixed in 8f3eb4389cbb5ee08e11913074d84319e8354e2f and 03042ee37cad233345b674f97089fc1638111036

File: crates/apollo-router/src/apollo_router.rs
Comment (o0Ignition0o) [reply crates/apollo-router/src/apollo_router.rs]: Ok i didn't know of meaning number 2., it makes sense!

I'm all for merging this as well, Its much better than it was before, and if I understand it correctly you meant this PR as a first step towards a more general refactoring.

+1 for me as well

**Comment by cecton** [comment]
@Geal @o0Ignition0o @BrynCooke let's be done with the discussion. We heard everybody. Time to vote. Select the rename you want to see:

1. No rename at all
2. create_route -> prepare
3. Route -> PreparedQuery
4. create_route -> prepare_query
5. Router -> Orchastrator 
6. Router -> Scheduler
7. Router -> QueryPlanner
8. Route -> Executor

Comments are editable so feel free to add elements to the list in this comment. You can select as many as you want. Majority wins.

I vote for 1.

**Comment by o0Ignition0o** [comment]
I like 3, but it's a weakly held opinion

File: crates/apollo-router/src/apollo_router.rs
Comment (abernix) [reply crates/apollo-router/src/apollo_router.rs]: Please tag me into that convo!  Happy to discuss. 😉

**Comment by BrynCooke** [comment]
I like 3, with any combination of 2 or 4.

**Comment by Geal** [comment]
I wish we took a little more time to think about it, but I guess it would become easier to think about once all the changes are done.

I'm wary of PreparedQuery, since it's close to [prepared statements](https://en.wikipedia.org/wiki/Prepared_statement). I think in general we should look at database vocabulary, since they perform similar tasks and already found names for them.

I'm against QueryPlanner since we already have a query planner component somewhere else, unless we find a way to merge them.

I'd say 2, 5, 6, 8

**Comment by BrynCooke** [comment]
> I think in general we should look at database vocabulary, since they perform similar tasks and already found names for them.

This is exactly why PreparedQuery is proposed. It's similar to database vocabulary and is a pattern that is widely used.

Agree -1 to QueryPlanner.
Stuff like Executor and Scheduler I would associate with threading, but that's just because I'm a Java dev. Maybe this is not the case in Rust.

Orchestration would be OK.

**Comment by cecton** [comment]
Ok so this is the result:
- Item 3: Route -> PreparedQuery
- Item 4: create_route -> prepare_query

Router will not be renamed because there is no majority.

**Comment by cecton** [comment]
@Geal @o0Ignition0o @BrynCooke awaiting for you approval :hugs:

**Review by o0Ignition0o — APPROVED** [review]
go for it!

**Review by Geal — APPROVED** [review]

**Pull request merged**
